AMQP 1.0 Transporter

vladir95 opened the pull request:
## :memo: Description

Added moleculer transporter for AMQP 1.0 protocol.

### :dart: Relevant issues
#561

### :gem: Type of change

- [X] This change requires a documentation update
- [X] New feature (non-breaking change which adds functionality)

### :scroll: Example code
```js
	const broker = new ServiceBroker({
		nodeID: "node",
		logger: false,
		transporter: "amqp10://admin:admin@localhost:5672"
	});
``` 

## :vertical_traffic_light: How Has This Been Tested?

Tested with moleculer 0.13.11 against ActiveMQ broker 5.15.10
- [X] Unit Tests
- [X] Flow Tests
- [X] Some load tests

## :checkered_flag: Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] **I have added tests that prove my fix is effective or that my feature works**
- [X] **New and existing unit tests pass locally with my changes**
- [X] I have commented my code, particularly in hard-to-understand areas

icebob commented:
Awesome! I'm going to check it.

AndreMaz commented:
Hey @vladir95 thanks for your work.
I've tried to do some simple tests over here but the nodes keep disconnecting due to timeout
```bash
[2019-12-20T11:54:37.156Z] INFO  node-2/BROKER: Transporter: Amqp10Transporter
[2019-12-20T11:54:37.156Z] INFO  node-2/BROKER: Registered 10 internal middleware(s).
[2019-12-20T11:54:37.159Z] INFO  node-1/TRANSIT: Connecting to the transporter...
[2019-12-20T11:54:37.179Z] INFO  node-2/TRANSIT: Connecting to the transporter...
[2019-12-20T11:54:37.184Z] INFO  node-2/TRANSPORTER: AMQP10 is connected.
[2019-12-20T11:54:37.188Z] INFO  node-1/TRANSPORTER: AMQP10 is connected.

[2019-12-20T11:55:37.207Z] INFO  node-2/TRANSPORTER: AMQP10 is disconnected.
[2019-12-20T11:55:37.209Z] ERROR node-2/TRANSPORTER: OperationTimeoutError: Unable to create the amqp receiver 'topic://MOL.DISCOVER' on amqp session 'local-3_remote-3_connection-2' due to operation timeout.
    at Timeout.actionAfterTimeout (/home/dev/tests/node_modules/rhea-promise/dist/lib/session.js:268:31)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7) {
  name: 'OperationTimeoutError'
}
[2019-12-20T11:55:37.209Z] WARN  node-2/TRANSIT: Connection is failed. Unable to create the amqp receiver 'topic://MOL.DISCOVER' on amqp session 'local-3_remote-3_connection-2' due to operation timeout.
[2019-12-20T11:55:37.209Z] INFO  node-1/TRANSPORTER: AMQP10 is disconnected.
```

Steps to repro: Just connect to `transporter: "amqp10://localhost:5672"` and wait like 30 seconds. Also, I'm using RabbitMQ 3.8.2 with default configs

I've also got (once) TypeError
```bash
ERROR dev/TRANSPORTER: TypeError: Cannot read property 'close' of null
    at Amqp10Transporter.disconnect (/home/dev/tests/node_modules/moleculer/src/transporters/amqp10.js:230:27)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

Can you check what's happening?

AndreMaz commented on src/transporters/amqp10.js:
```diff
@@ -0,0 +1,456 @@
+		this.connection = null;
+	}
+
+	_getQueueOptions(packetType, balancedQueue) {
```
Due to many `ToDos` can you explain a little bit more about what's missing?

vladir95 commented:
> Hey @vladir95 thanks for your work.
> I've tried to do some simple tests over here but the nodes keep disconnecting due to timeout
> 
> ```shell
> [2019-12-20T11:54:37.156Z] INFO  node-2/BROKER: Transporter: Amqp10Transporter
> [2019-12-20T11:54:37.156Z] INFO  node-2/BROKER: Registered 10 internal middleware(s).
> [2019-12-20T11:54:37.159Z] INFO  node-1/TRANSIT: Connecting to the transporter...
> [2019-12-20T11:54:37.179Z] INFO  node-2/TRANSIT: Connecting to the transporter...
> [2019-12-20T11:54:37.184Z] INFO  node-2/TRANSPORTER: AMQP10 is connected.
> [2019-12-20T11:54:37.188Z] INFO  node-1/TRANSPORTER: AMQP10 is connected.
> 
> [2019-12-20T11:55:37.207Z] INFO  node-2/TRANSPORTER: AMQP10 is disconnected.
> [2019-12-20T11:55:37.209Z] ERROR node-2/TRANSPORTER: OperationTimeoutError: Unable to create the amqp receiver 'topic://MOL.DISCOVER' on amqp session 'local-3_remote-3_connection-2' due to operation timeout.
>     at Timeout.actionAfterTimeout (/home/dev/tests/node_modules/rhea-promise/dist/lib/session.js:268:31)
>     at listOnTimeout (internal/timers.js:531:17)
>     at processTimers (internal/timers.js:475:7) {
>   name: 'OperationTimeoutError'
> }
> [2019-12-20T11:55:37.209Z] WARN  node-2/TRANSIT: Connection is failed. Unable to create the amqp receiver 'topic://MOL.DISCOVER' on amqp session 'local-3_remote-3_connection-2' due to operation timeout.
> [2019-12-20T11:55:37.209Z] INFO  node-1/TRANSPORTER: AMQP10 is disconnected.
> ```
> 
> Steps to repro: Just connect to `transporter: "amqp10://localhost:5672"` and wait like 30 seconds. Also, I'm using RabbitMQ 3.8.2 with default configs
> 
> I've also got (once) TypeError
> 
> ```shell
> ERROR dev/TRANSPORTER: TypeError: Cannot read property 'close' of null
>     at Amqp10Transporter.disconnect (/home/dev/tests/node_modules/moleculer/src/transporters/amqp10.js:230:27)
>     at runMicrotasks (<anonymous>)
>     at processTicksAndRejections (internal/process/task_queues.js:93:5)
> ```
> 
> Can you check what's happening?

RabbitMQ has no support out of the box for AMQP1.0 protocol, did you use the additional plugin?

AndreMaz commented:
> RabbitMQ has no support out of the box for AMQP1.0 protocol, did you use the additional plugin?

Yes, I've used the plugin (also with default configs) and I did manage to connect. But after some time nodes disconnected

AndreMaz commented on src/transporters/amqp10.js:
```diff
@@ -0,0 +1,456 @@
+			case PACKET_PING:
+			case PACKET_PONG:
+				// TODO: auto delete
+				break;
```
Due to many ToDos can you explain a little bit more about what's missing?

vladir95 replied:
we didn't find any naive `autoDeleteQueues` option implementation, so we left it as a `TODO`. anyways, you could always just overwrite `opts.queueOptions` or `opts.topicOptions` to your liking.

vladir95 commented:
> > RabbitMQ has no support out of the box for AMQP1.0 protocol, did you use the additional plugin?
> 
> Yes, I've used the plugin (also with default configs) and I did manage to connect. But after some time the nodes disconnected

Yes apparently RabbitMQ implements prefixes for queues and topics differently, https://github.com/rabbitmq/rabbitmq-amqp1.0/issues/18
We will look into that.

vladir95 commented:
@AndreMaz we added a new option called `topicPrefix`, it defaults to `topic://` which is more widely used, but you can override it, for rabbitmq you would set `/topic/`.

vladir95 commented:
Hey guys, did you have a chance to test it again?

icebob commented:
Yes, we will check it in the next days.

AndreMaz commented:
Hey @vladir95 sorry for the delay. 

> we added a new option called topicPrefix, it defaults to topic:// which is more widely used, but you can override it, for rabbitmq you would set /topic/.

Yes, this solved the issue. The nodes are no longer disconnecting :+1: 

---

I've been doing some testing and I think that there something wrong during file streaming. Thing are working fine with NATS but with AMQP10 the file is broken.

Can you please investigate what's happening?

Here is the repro code:
```js
const ServiceBroker = require("../src/service-broker");
const fs = require("fs");
const Promise = require("bluebird");

const transporter = {
	type: "AMQP10",
	options: {
		url: "amqp10://localhost:5672",
		topicPrefix: "/topic/"
	}
};

// Create broker #1
const broker1 = new ServiceBroker({
	nodeID: "client-" + process.pid,
	// transporter: "NATS",
	transporter,
	logLevel: "info"
});

// Create broker #2
const broker2 = new ServiceBroker({
	nodeID: "storage-" + process.pid,
	// transporter: "NATS",
	transporter,
	logLevel: "info"
});

broker2.createService({
	name: "storage",
	actions: {
		store(ctx) {
			return new Promise((resolve, reject) => {
				const stream = fs.createWriteStream(`dev/b.jpg`);

				stream.on("close", () => resolve("Done"));
				stream.on("error", err => reject(err));

				ctx.params.pipe(stream);
			});
		}
	}
});

broker1.createService({
	name: "sender",

	actions: {
		async send(ctx) {
			const fileName = "dev/a.jpg";
			const stream = fs.createReadStream(fileName);
			return await ctx.call("storage.store", stream);
		}
	}
});

Promise.all([broker1.start(), broker2.start()])
	.delay(2000)
	.then(async () => {
		broker1.repl();

		try {
			const result = await broker1.call("sender.send");

			broker1.logger.info(result);
		} catch (error) {
			broker1.logger.error(error);
		}
	});
```

vladir95 commented:
@icebob Thanks for the review, me and @nllahat will be testing and fixing those issues next week.

vladir95 commented on src/transporters/index.js:
```diff
@@ -53,6 +54,8 @@ function resolve(opt) {
 			TransporterClass = Transporters.Redis;
 		else if (opt.startsWith("amqp://") || opt.startsWith("amqps://"))
 			TransporterClass = Transporters.AMQP;
+		else if (opt.startsWith("amqp10://"))
```
No, you can pass secure connection settings to as `connectionOptions` to the transporter, we can allow something like `amqps10` that you can configure.

vladir95 commented on src/transporters/amqp10.js:
```diff
@@ -0,0 +1,460 @@
+
+			receiver.on("message", async context => {
+				await this._consumeCB(cmd, needAck)(context);
+				receiver.addCredit(1);
```
`this.opts.prefetch` defines how many simultaneous messages will be processed at once, each processed message takes 1 credit. So when it finishes we are re-adding the used credit.

vladir95 commented:
@AndreMaz, sorry for taking so long, we were busy :(

We looked into your issue, and it seems that the problem was with the prefetch window - how many concurrent messages are processed together.
From what i can see (correct me if i'm wrong), if you send a file stream, it sends many packets. The default prefetch window, is 1, meaning process 1 message at a time, but it seems that moleculer does not ack the delivery of the messages as they come in, but acks all the messages after they are all received.
That means that the first message is 'stuck', and no other messages are received, resulting in a nasty timeout.

You can see for yourself by changing `options.prefetch: 0` (after my recent fix to allow 0), that it'll work - because all the messages are received together.

It does not seem optimal. We saw in the tests, that moleculer by default, expects to process 1 message at a time, if no prefetch is given (in other transporters).

@icebob Would also like your input on that.

icebob commented on src/transporters/amqp10.js:
```diff
@@ -0,0 +1,498 @@
+		let receiverOptions = this._getQueueOptions(cmd);
+
+		if (nodeID) {
+			const needAck = [PACKET_REQUEST].indexOf(cmd) !== -1;
```
It's OK. In the future, we will add event acknowledgement into Moleculer core, as well.

vladir95 commented:
> Thanks taking the time to make this!
> 
> I found one issue where if I start a moleculer broker and then restart my RabbitMQ instance it will reconnect, but the queues and exchanges are gone and aren't recreated.

@Nathan-Schwartz Oh sorry about that, we removed an event listener by mistake with the last promises refactor :)
Fixed on last commit.

vladir95 commented on src/transporters/amqp10.js:
```diff
@@ -0,0 +1,498 @@
+							if (this.connection) {
+								delivery.reject();
+							}
+						});
```
For now we can change `delivery.reject();` to 'delivery.release();' to match the 0.9.1 transporter, but FYI, it will not pass the tests, as it expects the message to be back in the queue and retryable by other nodes.

Nathan-Schwartz replied:
Sorry for my slow response. Let's do `accept` in both. It might require changes in both test suites which I can help with if you'd like.

icebob commented:
@vladir95 @Nathan-Schwartz what's the status of it?

vladir95 commented:
@icebob  @Nathan-Schwartz I think it's pretty much done from our side, unless you find any other bugs or issues to address.

Nathan-Schwartz commented:
@vladir95 I tested it locally again and I found that with amqp10 I only get a response to the first action call. Subsequent calls run the action handler but no response is returned.

I verified this behavior exists with balancer enabled and disabled.

My code can be found here: https://gist.github.com/Nathan-Schwartz/8b04c9a12c3a2e4c123ab210ce6e2b98

nllahat commented:
> @vladir95 I tested it locally again and I found that with amqp10 I only get a response to the first action call. Subsequent calls run the action handler but no response is returned.
> 
> I verified this behavior exists with balancer enabled and disabled.
> 
> My code can be found here: https://gist.github.com/Nathan-Schwartz/8b04c9a12c3a2e4c123ab210ce6e2b98

Thanks @Nathan-Schwartz 
We will look into it ASAP

vladir95 commented:
@Nathan-Schwartz We fixed this issue as well, it was due to incorrect session management on our side.
Now everything uses the same session, instead of creating a new one each message, which caused the issue you were experiencing.

Nathan-Schwartz commented:
Thanks for the quick response!

I tested again and unfortunately I'm still getting the issue where I'm only getting one response.

I'm also getting roughly when the 7th message is sent:
> (node:1748) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 session_error listeners added. Use emitter.setMaxListeners() to increase limit

nllahat commented:
Hi @Nathan-Schwartz 

We've tested the code both with activeMQ and rabbitMQ(amqp10 plugin) and it worked ok. 
Did you ran the brokers after cleaning all queues and exchangers ? if not - can you try that and let us know if that helped ? 
In addition - we will be happy to know what rabbitMQ version are you using ? 


Thanks !!

nllahat commented:
@Nathan-Schwartz 

> (node:1748) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 session_error listeners added. Use emitter.setMaxListeners() to increase limit

We see this warning also. Will fix it (:

Nathan-Schwartz commented:
@nllahat Sorry about that! I forgot I had modified the transporter during my previous tests. Actions and events are working well for me now 👍 

This is pretty minor, but I noticed when I hit `Ctrl-C` the server throws an error. I don't think it is a big problem, but it doesn't happen with the other AMQP transporter.
```
events.js:189
    throw err; // Unhandled 'error' event
    ^

Error [ERR_UNHANDLED_ERROR]: Unhandled error. ({ _context:
   Error: transfer after detach
       at Incoming.on_transfer (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/session.js:364:15)
       at Session.on_transfer (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/session.js:749:19)
       at Connection.(anonymous function) [as on_transfer] (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/connection.js:786:30)
       at c.dispatch (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/types.js:910:33)
       at Transport.read (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/transport.js:109:36)
       at Connection.input (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/connection.js:520:35)
       at Socket.emit (events.js:198:13)
       at addChunk (_stream_readable.js:288:12)
       at readableAddChunk (_stream_readable.js:269:11)
       at Socket.Readable.push (_stream_readable.js:224:10)
       at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17),
```

I'm also noticing that events get acked before the handler completes, so messages are all processed right away. For actions they queue up until the handler finishes. This behavior is present for 0-9-1 as well so I don't think it should block this PR but I wanted to mention it.

vladir95 commented:
> @nllahat Sorry about that! I forgot I had modified the transporter during my previous tests. Actions and events are working well for me now 👍
> 
> This is pretty minor, but I noticed when I hit `Ctrl-C` the server throws an error. I don't think it is a big problem, but it doesn't happen with the other AMQP transporter.
> 
> ```
> events.js:189
>     throw err; // Unhandled 'error' event
>     ^
> 
> Error [ERR_UNHANDLED_ERROR]: Unhandled error. ({ _context:
>    Error: transfer after detach
>        at Incoming.on_transfer (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/session.js:364:15)
>        at Session.on_transfer (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/session.js:749:19)
>        at Connection.(anonymous function) [as on_transfer] (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/connection.js:786:30)
>        at c.dispatch (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/types.js:910:33)
>        at Transport.read (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/transport.js:109:36)
>        at Connection.input (/Users/nathan.schwartz/Documents/projects/moleculer/node_modules/rhea/lib/connection.js:520:35)
>        at Socket.emit (events.js:198:13)
>        at addChunk (_stream_readable.js:288:12)
>        at readableAddChunk (_stream_readable.js:269:11)
>        at Socket.Readable.push (_stream_readable.js:224:10)
>        at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17),
> ```
> 
> I'm also noticing that events get acked before the handler completes, so messages are all processed right away. For actions they queue up until the handler finishes. This behavior is present for 0-9-1 as well so I don't think it should block this PR but I wanted to mention it.

Yeah that seems more related to `rhea` + `rhea-promise`, as for the second thing - yeah we mimicked the 0-9-1 behaviour there, but it can be easily changed in future versions, just need to add `needAck` for event packets, currently they are only for requests.

icebob commented:
@Nathan-Schwartz so can we merge it?
Just I'm asking because I would like to release 0.14 at the weekend and would be good to merge it into 0.14, as well before release. Is it possible?

icebob commented:
@vladir95 do you need to be published in 0.13 or enough in 0.14?

vladir95 commented:
> @vladir95 do you need to be published in 0.13 or enough in 0.14?

@icebob we need it in 0.13 also, thanks :)

icebob commented:
OK, could you help with the documentation, as well. Would be good to add this new transporter to this page: https://moleculer.services/docs/0.13/networking.html#Transporters
Source file: https://github.com/moleculerjs/site/blob/master/source/docs/0.13/networking.md

Thanks in advance!

vladir95 commented:
> https://github.com/moleculerjs/site/blob/master/source/docs/0.13/networking.md

@icebob  https://github.com/moleculerjs/site/pull/92 🙂

icebob commented:
Cool, thank you very much!

icebob commented:
It's released in 0.13.13

icebob commented:
@vladir95 Hi,
I've made some integration tests with Docker & Kubernetes and it generates & starts a simple Moleculer project with all transporters. All are good, except the AMQP1.0 transporter. And I don't understand what is the problem. Could you help me to check, what is the problem with it?

Here is the log of ActiveMQ container:
https://github.com/moleculerjs/moleculer-template-project/runs/656329054?check_suite_focus=true#step:13:45

vladir95 commented:
> @vladir95 Hi,
> I've made some integration tests with Docker & Kubernetes and it generates & starts a simple Moleculer project with all transporters. All are good, except the AMQP1.0 transporter. And I don't understand what is the problem. Could you help me to check, what is the problem with it?
> 
> Here is the log of ActiveMQ container:
> https://github.com/moleculerjs/moleculer-template-project/runs/656329054?check_suite_focus=true#step:13:45

Sure.
Just making sure I understand your flow correctly - 
You are starting 3 containers in moleculer and an activemq one, trying to connect and it throws this error, no special configuration, just basic amqp1.0 configuration, right?
Can i have a link to the docker-compose you are using?

icebob commented:
Yeah, sure.

It runs more containers, but only one transporter:
```
       Name                     Command               State                               Ports                            
---------------------------------------------------------------------------------------------------------------------------
ci-test_activemq_1   /bin/sh -c bin/activemq co ...   Up      1883/tcp, 5672/tcp, 61613/tcp, 61614/tcp, 61616/tcp, 8161/tcp
ci-test_api_1        docker-entrypoint.sh npm start   Up                                                                   
ci-test_greeter_1    docker-entrypoint.sh npm start   Up                                                                   
ci-test_mongo_1      docker-entrypoint.sh mongod      Up      27017/tcp                                                    
ci-test_products_1   docker-entrypoint.sh npm start   Up                                                                   
ci-test_redis_1      docker-entrypoint.sh redis ...   Up      6379/tcp                                                     
ci-test_traefik_1    /entrypoint.sh --api.insec ...   Up      0.0.0.0:3000->80/tcp, 0.0.0.0:3001->8080/tcp           
```
Here is the ActiveMQ container in [compose](https://github.com/moleculerjs/moleculer-template-project/blob/master/template/docker-compose.yml):
```yml
  activemq:
    image: rmohr/activemq
    networks:
      - internal
```
And ServiceBroker connects to them as `TRANSPORTER=amqp10://guest:guest@activemq:5672`

vladir95 commented:
@icebob Sorry didnt have much time last week, can you open an issue about that so we can follow up there more easily instead of here?
I will look into that this week, I seem to recall running into something like that...

icebob commented:
Thanks @vladir95, here is the new issue: https://github.com/moleculerjs/moleculer/issues/733
